fe/fe-core/src/main/java/com/starrocks/catalog/TabletInvertedIndex.java

```diff
@@ -311,6 +305,136 @@ public void tabletReport(long backendId, Map<Long, TTablet> backendTablets,
                 tabletMigrationMap.size(), transactionsToClear.size(), transactionsToPublish.size(), (end - start));
     }
 
+    private void deleteTabletByConsistencyChecker(long tabletId, long backendId,
+                                                  String reason, Set<Long> invalidTablets) {
+        LOG.info("delete tablet {} on backend {} from inverted index by consistency checker, because: {}",
+                tabletId, backendId, reason);
+        deleteTablet(tabletId);
+        invalidTablets.add(tabletId);
+    }
+
+    /**
+     * Check the consistency between {@link com.starrocks.catalog.TabletInvertedIndex} and
+     * {@link com.starrocks.server.LocalMetastore}.
+     * <p>
+     * If we find an invalid tablet, i.e. it's neither in current catalog nor in recycle bin,
+     * we will remove it from {@link com.starrocks.catalog.TabletInvertedIndex} directly.
+     * And this process will also output a report in `fe.log`, including valid number of
+     * tablet and number of tablet in recycle bin for each backend.
+     */
+    public void checkTabletMetaConsistency() {
+        LocalMetastore localMetastore = GlobalStateMgr.getCurrentState().getLocalMetastore();
+        CatalogRecycleBin recycleBin = GlobalStateMgr.getCurrentRecycleBin();
+
+        Set<Long> invalidTablets = new HashSet<>();
+        // backend id -> <num of currently existed tablet, num of tablet in recycle bin>
+        Map<Long, Pair<Long, Long>> backendTabletNumReport = new HashMap<>();
+        List<Long> backendIds = GlobalStateMgr.getCurrentSystemInfo().getBackendIds();
+
+        long startTime = System.currentTimeMillis();
+        long scannedTabletCount = 0;
+
+        for (Long backendId : backendIds) {
+            List<Long> tabletIds = getTabletIdsByBackendId(backendId);
+            backendTabletNumReport.put(backendId, new Pair<>(0L, 0L));
+
+            for (Long tabletId : tabletIds) {
+                scannedTabletCount++;
+                boolean isInRecycleBin = false;
+
+                TabletMeta tabletMeta = getTabletMeta(tabletId);
+                if (tabletMeta == null) {
+                    deleteTabletByConsistencyChecker(tabletId, backendId, "tablet meta is null", invalidTablets);
+                    continue;
+                }
+
+                // validate database
+                long dbId = tabletMeta.getDbId();
+                Database database = localMetastore.getDb(dbId);
+                if (database == null) {
+                    database = recycleBin.getDatabase(dbId);
+                    if (database != null) {
+                        isInRecycleBin = true;
+                    } else {
+                        deleteTabletByConsistencyChecker(tabletId, backendId,
+                                "database " + dbId + " doesn't exist", invalidTablets);
+                        continue;
+                    }
+                }
+
+                try {
+                    database.readLock();
```

Comment:
For every tablet, we will get database lock once, how about group all the tablet in one database, and check them in batch?

Reply:
this's intended to do so, to avoid holding db lock for too long time and affect query or loading, the checking process here is background and don't need performance.